lib/ssl/src/tls_gen_connection.erl

```diff
@@ -77,36 +77,67 @@
 %% Setup
 %%====================================================================
 start_fsm(Role, Host, Port, Socket,
-          {#{erl_dist := false, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
+          {#{erl_dist := false = ErlDist, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
 	  User, {CbModule, _, _, _, _} = CbInfo, 
-	  Timeout) -> 
-    try
-        {ok, DynSup} = tls_connection_sup:start_child([]),
-        {ok, Sender} = tls_dyn_connection_sup:start_child(DynSup, sender, [[{spawn_opt, SenderOpts}]]),
-	{ok, Pid} = tls_dyn_connection_sup:start_child(DynSup, receiver, [Role, Sender, Host, Port, Socket,
-                                                             Opts, User, CbInfo]),
-	{ok, SslSocket} = ssl_gen_statem:socket_control(?MODULE, Socket, [Pid, Sender], CbModule, Trackers),
-        ssl_gen_statem:handshake(SslSocket, Timeout)
-    catch
-	error:{badmatch, {error, _} = Error} ->
-	    Error
+	  Timeout) ->
+    Starter = start_connection_tree(User, ErlDist, [[{spawn_opt, SenderOpts}]],
+                                    Role, [Host, Port, Socket, Opts, User, CbInfo]),
+    receive
+        {Starter, SockReceiver, SockSender} ->
+            socket_control(Socket, SockReceiver, SockSender, CbModule, Trackers, Timeout);
+        {Starter, Error} ->
+            Error
     end;
-
 start_fsm(Role, Host, Port, Socket,
-          {#{erl_dist := true, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
+          {#{erl_dist := true = ErlDist, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
 	  User, {CbModule, _, _, _, _} = CbInfo, 
-	  Timeout) -> 
-    try
+	  Timeout) ->
         SenderOpts1 = [{priority, max} | proplists:delete(priority, SenderOpts)],
```

Comment:
Since the 2 start_fsm clauses are identical except for one line, I propose that they are put together like this:
```erlang
start_fsm(Role, Host, Port, Socket,
          {#{erl_dist := true = ErlDist, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
	  User, {CbModule, _, _, _, _} = CbInfo, 
	  Timeout) ->
        SenderOpts1 = set_erl_dist_priority(ErlDist,SenderOpts),
    Starter = start_connection_tree(User, ErlDist, [[{spawn_opt, SenderOpts1}]],
                                    Role, [Host, Port, Socket,Opts, User, CbInfo]),
    receive
        {Starter, SockReceiver, SockSender} ->
            socket_control(Socket, SockReceiver, SockSender, CbModule, Trackers, Timeout);
        {Starter, Error} ->
            Error
    end.

%% put this helper function in a suitable place in the module
set_erldist_priority(true,SenderOpts) ->
        [{priority, max} | proplists:delete(priority, SenderOpts)];
set_erldist_priority(false,SenderOpts) ->
       SenderOpts.

```

Reply:
Sure we can remove some code duplication here. I prefer to generalize the helper function to handle the sender options.